oio/cli/admin/directory.py

```diff
@@ -175,3 +178,94 @@ def take_action(self, parsed_args):
         mapping.load(connection_timeout=10.0, read_timeout=60.0)
         mapping.decommission(parsed_args.addr)
         mapping.force(connection_timeout=10.0, read_timeout=60.0)
+
+
+class DirectoryWarmup(DirectoryCmd):
+    """Ping each prefix of a Meta0 hash to prepare each Meta1 base"""
+
+    def get_parser(self, prog_name):
+        parser = super(DirectoryCmd, self).get_parser(prog_name)
+        parser.add_argument('--workers', type=int, default=1,
+                            help="How many concurrent bases to warm up")
+        parser.add_argument('--proxy', type=str, default=None,
+                            help="Specific proxy IP:PORT")
+        return parser
+
+    def _ping_prefix(self, prefix):
+        pass
+
+    def take_action(self, parsed_args):
+        self.log.debug('take_action(%s)', parsed_args)
+        digits = self.app.client_manager.get_meta1_digits()
+        workers_count = parsed_args.workers
+
+        conf = {'namespace': self.app.client_manager.namespace}
+        if parsed_args.proxy:
+            conf.update({'proxyd_url': parsed_args.proxy})
+        else:
+            ns_conf = load_namespace_conf(conf['namespace'])
+            proxy = ns_conf.get('proxy')
+            conf.update({'proxyd_url': proxy})
+
+        workers = list()
+        with green.ContextPool(workers_count) as pool:
+            pile = GreenPile(pool)
+            prefix_queue = Queue(16)
+
+            # Prepare some workers
+            for i in range(workers_count):
+                w = WarmupWorker(conf, self.log)
+                workers.append(w)
+                pile.spawn(w.run, prefix_queue)
+
+            # Feed the queue
+            trace_increment = 0.01
+            trace_next = trace_increment
+            sent, total = 0, float(count_prefixes(digits))
+            for prefix in generate_prefixes(digits):
+                sent += 1
+                prefix_queue.put(prefix)
+                # Display the progression
+                ratio = float(sent) / total
+                if ratio >= trace_next:
+                    self.log.info("... %d%%", int(ratio * 100.0))
+                    trace_next += trace_increment
+
+            self.log.debug("Send the termination marker")
+            for i in range(workers_count):
+                prefix_queue.put(None)
```

Comment:
You can use [`prefix_queue.task_done()`](https://docs.python.org/2/library/queue.html#Queue.Queue.task_done). And you wait for the end with [`prefix_queue.join()`](https://docs.python.org/2/library/queue.html#Queue.Queue.join).